pkg/dogstatsd/enrich.go

```diff
@@ -0,0 +1,199 @@
+package dogstatsd
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/tagger/collectors"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type tagRetriever func(entity string, cardinality collectors.TagCardinality) ([]string, error)
+
+var (
+	hostTagPrefix     = "host:"
+	entityIDTagPrefix = "dd.internal.entity_id:"
+
+	getTags tagRetriever = tagger.Tag
+)
+
+func parseMetricMessage(message []byte, namespace string, namespaceBlacklist []string, defaultHostname string) (*metrics.MetricSample, error) {
+	sample, err := parseMetricSample(message)
+	if err != nil {
+		return nil, err
+	}
+	return enrichMetricSample(sample, namespace, namespaceBlacklist, defaultHostname), nil
+}
+
+func parseEventMessage(message []byte, defaultHostname string) (*metrics.Event, error) {
+	sample, err := parseEvent(message)
+	if err != nil {
+		return nil, err
+	}
+	return enrichEvent(sample, defaultHostname), nil
+}
+
+func parseServiceCheckMessage(message []byte, defaultHostname string) (*metrics.ServiceCheck, error) {
+	sample, err := parseServiceCheck(message)
+	if err != nil {
+		return nil, err
+	}
+	return enrichServiceCheck(sample, defaultHostname), nil
+}
+
+func enrichTags(tags []string, defaultHostname string) ([]string, string) {
+	if len(tags) == 0 {
+		return nil, defaultHostname
+	}
+
+	extraTags := make([]string, 0, 8)
+	host := defaultHostname
+
+	n := 0
+	for _, tag := range tags {
+		if strings.HasPrefix(tag, hostTagPrefix) {
+			host = string(tag[len(hostTagPrefix):])
+		} else if strings.HasPrefix(tag, entityIDTagPrefix) {
+			// currently only supported for pods
+			entity := kubelet.KubePodTaggerEntityPrefix + string(tag[len(entityIDTagPrefix):])
+			entityTags, err := getTags(entity, tagger.DogstatsdCardinality)
+			if err != nil {
+				log.Tracef("Cannot get tags for entity %s: %s", entity, err)
+				continue
+			}
+			extraTags = append(extraTags, entityTags...)
+		} else {
+			tags[n] = tag
+			n++
+		}
+	}
```

Comment:
In case there is no enrichment should we just return here? To avoid the two following operations:

```
	tags = tags[:n]
	tags = append(tags, extraTags...)
```

Reply:
IMO this would be premature optimization at the cost of some readability. I try to force myself to measure first before taking any decision that trades something for hypothetic performance improvements. 

It's very likely that there are fast path for those cases in Go's runtime and that the overhead is negligible.